pkg/controller/reconciler/recipe_webhook_test.go

```diff
@@ -0,0 +1,330 @@
+		recipeName := "test-recipe-invalidtype"
+		namespace := types.NamespacedName{Namespace: defaultNamespace, Name: recipeName}
+		recipe := makeRecipe(namespace, invalidResourceType)
+
+		err := client.Create(ctx, &corev1.Namespace{ObjectMeta: ctrl.ObjectMeta{Name: namespace.Name}})
+		require.NoError(t, err)
+
+		// Webhook is expected to trigger during this call and return an error.
+		err = client.Create(ctx, recipe)
+		require.True(t, apierrors.IsForbidden(err))
+	})
+
+	t.Run("test recipe for valid type", func(t *testing.T) {
+		recipeName := "test-recipe-validtype"
+		namespace := types.NamespacedName{Namespace: defaultNamespace, Name: recipeName}
+		recipe := makeRecipe(namespace, validResourceType)
+
+		err := client.Create(ctx, &corev1.Namespace{ObjectMeta: ctrl.ObjectMeta{Name: namespace.Name}})
+		require.NoError(t, err)
+
+		err = client.Create(ctx, recipe)
+		require.NoError(t, err)
+
+		// Recipe will be waiting for extender to complete provisioning.
+		status := waitForRecipeStateUpdating(t, client, namespace, nil)
+
+		radius.CompleteOperation(status.Operation.ResumeToken, nil)
+		_, err = radius.Resources(status.Scope, validResourceType).Get(ctx, namespace.Name)
+		require.NoError(t, err)
+
+		err = client.Delete(ctx, recipe)
+		require.NoError(t, err)
+
+		// Deletion of the resource is in progress.
+		status = waitForRecipeStateDeleting(t, client, namespace, nil)
+		radius.CompleteOperation(status.Operation.ResumeToken, nil)
+
+		// Now deleting of the deployment object can complete.
+		waitForRecipeDeleted(t, client, namespace)
+
+	})
+
+	t.Run("test recipe update from valid to invalid type", func(t *testing.T) {
+		// Create a recipe with a valid type
+		recipeName := "test-recipe-update"
+		namespace := types.NamespacedName{Namespace: defaultNamespace, Name: recipeName}
+		recipe := makeRecipe(namespace, validResourceType)
+
+		err := client.Create(ctx, &corev1.Namespace{ObjectMeta: ctrl.ObjectMeta{Name: namespace.Name}})
+		require.NoError(t, err)
+
+		err = client.Create(ctx, recipe)
+		require.NoError(t, err)
+
+		// Recipe will be waiting for extender to complete provisioning.
+		status := waitForRecipeStateUpdating(t, client, namespace, nil)
+
+		radius.CompleteOperation(status.Operation.ResumeToken, nil)
+		_, err = radius.Resources(status.Scope, validResourceType).Get(ctx, namespace.Name)
+		require.NoError(t, err)
+
+		// Using RetryOnConflict to avoid catching a conflict error when updating the recipe.
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// Retrieve the latest version of the recipe
+			recipe = &radappiov1alpha3.Recipe{}
+			if err := client.Get(ctx, namespace, recipe); err != nil {
+				return err
+			}
+
+			// Update the recipe to have an invalid type
+			recipe.Spec.Type = invalidResourceType
+			return client.Update(ctx, recipe)
+		})
+		// The webhook should reject the update and return a Forbidden error
+		require.True(t, apierrors.IsForbidden(err))
```

Comment:
per dicussion, this isn't the right return code. We should return an `Invalid` error from the webhook so the right return code shows up here.

Reply:
updated pls take a look.